### ISSUE TITLE
invalidate cache on read error

### ISSUE THREAD

**Issue by gaillard** [jcupitt/libvips#432]
I have a situation where I use the C version of vips to resize an image but when I loaded the image it could have not been fully written. I eventually then come back and load it again and do the resize again but it is seeing the partial data from the first load.

 I think this has something to do with the operations cache? I am calling vips_thread_shutdown after my operations. How does vips identify if the cache is stale with relation to the file handle it has open?

Functions I am using are vips_jpegload, vips_autorot, vips_shrink, vips_affine, vips_jpegsave and vips_crop.

**Comment by jcupitt** [comment]
Hi @gaillard, it should all work, I've been trying to think of some way this could fail and I'm coming up blank.

Could you post a small program which shows this bad behaviour?

**Comment by gaillard** [comment]
I'll see if I can get a small test case going. I am going through CGo in case that makes a difference.

**Comment by jcupitt** [comment]
Hi, I read your question again more carefully, is this the sequence of events?

1. Start writing an image to file A on thread1
2. Before thread1 has finished writing, thread2 opens file A, reads some pixels, and encounters an error since the image is incomplete
3. Thread2 closes the image
4. Thread1 finishes writing file A
5. Thread3 tries to open file A, but it sees the broken image from step 2, vips will not reopen the image again.

Yes, you're right, this could happen because of the vips operation cache. The image open will succeed, because it'll just read the header, so the open image operation will go into the cache. The read error will occur later and won't force the broken image out of cache. 

I'm not sure what the best solution is. Perhaps if the file format readers hit an error they should kick themselves out of cache?

**Comment by gaillard** [comment]
Yes exactly. I am doing big bunches of batches so I don't really need the cache except when I set max files to zero it still seems to hold one open?

As to a solution to the root problem, besides watching files you could check source file ctimes when the files are opened in vips. You would have the resolution issue but the caller could ensure that.

**Comment by jcupitt** [comment]
I've had a quick hack here:

https://github.com/jcupitt/libvips/tree/invalidate-on-load-error

If there's an error from libjpeg during jpeg read, it signals "invalidate" on the `jpegload` operation, which knocks it out of the operation cache.

It seems to work, I think. Would you be able to test?

If it looks OK, we need to add a little extra code to all readers.

**Comment by jcupitt** [comment]
Here's a simple test program:

```python
#!/usr/bin/python

import sys
import time
import gi
gi.require_version('Vips', '8.0')
from gi.repository import Vips

while True:
    a = Vips.Image.new_from_file(sys.argv[1])
    print "a.avg() =", a.avg()
    time.sleep(1)
```

I ran it like this:

```
$ cp ~/pics/broken.jpg x.jpg
$ ./sleep.py x.jpg & 
$ 
a.avg() = 156.440825144
a.avg() = 156.440825144
... etc.
```

`broken.jpg` has been truncated, so you get the average of the pixels that are there. Now if I copy a working file on top of `x.jpg` the output changes, showing that the load operation was not cached:

```
...
a.avg() = 156.440825144
$ cp ~/pics/k1.jpg x.jpg
a.avg() = 222.397428725
a.avg() = 222.397428725
...
```

If I copy a different file again, output does not change, showing that the successful read of `k1.jpg` has now been held.

**Comment by jcupitt** [comment]
Hi again, did you get a chance to look at this?

**Comment by gaillard** [comment]
Sorry I haven't been able to get back to this, our code has changed a lot so I'll have to retest but it was a hard scenario to reproduce in the first place. We don't reuse paths and/or attempt to read partially written files anymore so its largely fixed for us but still an issue.

**Comment by jcupitt** [comment]
OK, interesting to know this is still an issue.

It is on the 8.5 roadmap, so it should be added soon.

**Comment by jcupitt** [comment]
OK, merged to master! This should be in 8.5, due in March 2017. It won't work for all file types, but the major ones are done. 

Thank you for suggesting this useful improvement.
